expyfun/analyze/_analyze.py

```diff
@@ -42,39 +55,67 @@ def press_times_to_hmfc(presses, targets, foils, tmin, tmax,
+    else:
+        singleton = False
+    for r in return_type:
+        if not isinstance(r, string_types) or r not in known_types:
+            raise ValueError('r must be one of %s, got %s' % (known_types, r))
     # Sanity check that targets and foils don't overlap (due to tmin/tmax)
     targets = np.atleast_1d(targets) + tmin
     foils = np.atleast_1d(foils) + tmin
     dur = float(tmax - tmin)
     assert dur > 0
     presses = np.sort(np.atleast_1d(presses))
     assert targets.ndim == foils.ndim == presses.ndim == 1
-    all_times = np.concatenate(([-np.inf], targets, foils, [np.inf]))
-    order = np.argsort(all_times)
-    inv_order = np.argsort(order)
-    all_times = all_times[order]
-    if not np.all(all_times[:-1] + dur <= all_times[1:]):
+    # Stack as targets, then foils
+    stim_times = np.concatenate(([-np.inf], targets, foils, [np.inf]))
+    order = np.argsort(stim_times)
+    stim_times = stim_times[order]
+    if not np.all(stim_times[:-1] + dur <= stim_times[1:]):
         raise ValueError('Analysis windows for targets and foils overlap')
-    # Let's just loop (could probably be done with vector math, but it's
-    # too hard and unlikely to be correct)
-    locs = np.searchsorted(all_times, presses, 'right')
-    if len(locs) > 0:
-        assert locs.max() < len(all_times)  # should be True b/c of np.inf
-        assert locs.min() >= 1
-
-    # figure out which presses were to target or masker (valid_idx)
-    in_window = (presses <= all_times[locs - 1] + dur)
-    valid_idx = np.where(in_window)[0]
-    n_other = np.sum(~in_window)
-
+    # figure out what targ/mask times our presses correspond to
+    press_to_stim = np.searchsorted(stim_times, presses) - 1
+    if len(press_to_stim) > 0:
+        assert press_to_stim.max() < len(stim_times)  # True b/c of np.inf
+        assert press_to_stim.min() >= 0
+    stim_times = stim_times[press_to_stim]
+    order = order[press_to_stim]
+    assert (stim_times <= presses).all()
+
+    # figure out which presses were to target or masker (valid_mask)
+    valid_mask = (presses <= stim_times + dur)
+    n_other = np.sum(~valid_mask)
+    press_to_stim = press_to_stim[valid_mask]
+    presses = presses[valid_mask]
+    stim_times = stim_times[valid_mask]
+    order = order[valid_mask]
+    del valid_mask
+    press_to_stim, used_map_idx = np.unique(press_to_stim, return_index=True)
+    presses = presses[used_map_idx]
+    stim_times = stim_times[used_map_idx]
+    order = order[used_map_idx]
+    assert len(presses) == len(stim_times)
+    diffs = presses - (stim_times - tmin)
+    assert all(diffs >= tmin)
+    assert all(diffs <= tmax)
+    del used_map_idx
     # figure out which of valid presses were to target or masker
-    used = np.unique(locs[valid_idx])  # unique to remove double-presses
-    orig_places = (inv_order[used - 1] - 1)
-    n_hit = sum(orig_places < len(targets))
-    n_fa = len(used) - n_hit
+    target_mask = (order <= len(targets))
```

Comment:
Maybe the most constructive thing to do is to come up with examples that you think will break the code and test them. It will be more productive than chatting about I think. You can make a PR into my branch with more test cases, passing or failing, if you want?